### PR TITLE
Fix serialization of custom types in caveat context

### LINKED ISSUE #1299: could not serialize context: proto: invalid type: types.IPAddress. please report this error
I am trying to develop an IP range check and I am using this as a base: https://authzed.com/docs/reference/caveats#ip-address-checking

Since I couldn't find a Playground example, I created one myself, based on your unit test https://github.com/authzed/spicedb/blob/main/pkg/caveats/types_test.go

Here is the Playground link: https://play.authzed.com/s/FGmHO5fqLhH8/assertions

When I try to run the assertion or generate expected relations, I get the following error:
`could not serialize context: proto: invalid type: types.IPAddress. please report this error`

I only need a working example in the Playground.

Cheers.

### PR DESCRIPTION
**Pull request by josephschorr**
We need to have another method on a shared interface to enable this to be generic and be enforced

Also adds some tests

Fixed #1299

### CODE REVIEW

File: pkg/caveats/types/registration.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ func registerGenericType(
 }
 
 // registerCustomType registers a custom type that wraps a base CEL type.
-func registerCustomType(keyword string, baseCelType *cel.Type, converter typedValueConverter, opts ...cel.EnvOption) VariableType {
+func registerCustomType[T CustomType](keyword string, baseCelType *cel.Type, converter typedValueConverter, opts ...cel.EnvOption) VariableType {
```
Comment (vroldanbet) [inline pkg/caveats/types/registration.go:83]: I believe this generic is not being used and can be dropped

File: pkg/caveats/types/ipaddress.go [hunk 2]
```diff
@@ -67,7 +71,7 @@ func (ipa IPAddress) Value() interface{} {
 	return ipa
 }
 
-var IPAddressType = registerCustomType(
+var IPAddressType = registerCustomType[IPAddress](
```
Comment (vroldanbet) [inline pkg/caveats/types/ipaddress.go:74]: the generic type here could be dropped too

**Review by vroldanbet — COMMENTED** [review]
LGTM, just left a suggestion

File: pkg/caveats/types/registration.go (hunk 1 above)
Comment (josephschorr) [reply pkg/caveats/types/registration.go:83]: It is explicitly added to type check against the interface, to ensure it meets the `CustomType` requirements

File: pkg/caveats/types/ipaddress.go (hunk 2 above)
Comment (josephschorr) [reply pkg/caveats/types/ipaddress.go:74]: See above

**Review by vroldanbet — APPROVED** [review]

**Pull request merged**